tiatoolbox/tools/patchextraction.py

```diff
@@ -0,0 +1,330 @@
+            patch_size=patch_size,
+            pad_size=pad_size,
+        )
+        self.stride = stride
+
+    # def extract_patches(self, input_img, labels=None):
+    #     raise NotImplementedError
+
+    def __next__(self):
+        raise NotImplementedError
+
+    def merge_patches(self, patches):
+        raise NotImplementedError
+
+
+class VariableWindowPatchExtractor(PatchExtractor):
+    """Extract and merge patches using variable sized windows for images and labels.
+
+    Args:
+        stride(Tuple of int): stride in (x, y) direction for patch extraction.
+        label_patch_size(Tuple of int): network output label (width, height).
+
+    Attributes:
+        stride(Tuple of int): stride in (x, y) direction for patch extraction.
+        label_patch_size(Tuple of int): network output label (width, height).
+
+    """
+
+    def __init__(
+        self,
+        input_img,
+        patch_size,
+        pad_size=0,
+        stride=1,
+        label_patch_size=None,
+    ):
+        super().__init__(
+            input_img=input_img,
+            patch_size=patch_size,
+            pad_size=pad_size,
+        )
+        self.stride_h = stride
+        self.label_patch_size = label_patch_size
+
+    # def extract_patches(self, input_img, labels=None):
+    #     raise NotImplementedError
+
+    def __next__(self):
+        raise NotImplementedError
+
+    def merge_patches(self, patches):
+        raise NotImplementedError
+
+
+class PointsPatchExtractor(PatchExtractor):
+    """Extracting patches with specified points as a centre.
+
+    Args:
+        labels(ndarray, pd.DataFrame, str, pathlib.Path): contains location and/or
+         type of patch. Input can be path to a csv or json file.
+        num_examples_per_patch(int): Number of examples per patch for ensemble
+         classification, default=9 (centre of patch and all the eight neighbours as
+         centre).
+
+    Attributes:
+        labels(pd.DataFrame): A table containing location and/or type of patch.
+        num_examples_per_patch(int): Number of examples per patch for ensemble
+         classification.
+
+    """
+
+    def __init__(
+        self,
+        input_img,
+        labels,
+        patch_size=224,
+        pad_size=0,
+        num_examples_per_patch=9,
+    ):
+        super().__init__(
+            input_img=input_img,
+            patch_size=patch_size,
+            pad_size=pad_size,
+        )
+
+        self.num_examples_per_patch = num_examples_per_patch
+        self.labels = read_point_annotations(input_table=labels)
+
+    def __next__(self):
+        n = self.n
+
+        if n >= self.labels.shape[0]:
+            raise StopIteration
+        self.n = n + 1
+        return self[n]
+
+    def __getitem__(self, item):
+        if type(item) is not int:
+            raise TypeError("Index should be an integer.")
+
+        if item >= self.labels.shape[0]:
+            raise IndexError
+
+        x, y, _ = self.labels.values[item, :]
```

Comment:
I guess it is important to return the classification label alongside the patch as it is necessary to have that data for classification, suggest using something like this:
`x, y, cls = self.labels.values[item, :]`
and in the end return like this:
`return data, cls`

Maybe even (x, y) information would be useful. Should we consider them as optional output?

Reply:
Also, if we want to support point lists with two columns, we should adjust this part accordingly.

Reply:
you are right but currently our focus is on inference so we don't need this yet. But this will be added when we add code for training models.